### PR TITLE
Updated CPhysicalJoin to derive the inner distribution in the case of tainted replicated (6X)

### PR DESCRIPTION
**Pull request by thedanhoffman**
This addresses https://github.com/greenplum-db/gpdb/issues/13058

Previously, CPhysicalJoin derived the outer distribution when it was tainted
replicated. It checked only for strict replicated and universal replicated and
returned the inner distribution in these cases (in this case, it satisfies
random). Tainted replicated wasn't considered and was causing an undercount (the
JOIN derived tainted replicated instead of random, which was causing the number
of columns to be undercounted, because it wrongly assumed that one segment
contained all output columns).

### CODE REVIEW

**Comment by chrishajas** [comment]
This is a backport from master right? Can we add that commit sha to the commit message. You can use `git cherry-pick -x <sha>` and it will automatically add that message too.

**Comment by thedanhoffman** [comment]
> This is a backport from master right? Can we add that commit sha to the commit message. You can use `git cherry-pick -x <sha>` and it will automatically add that message too.

Added the SHA hash of the cherry-picked commit. I also split the rpt regression test into ORCA and non-ORCA

**Comment by thedanhoffman** [comment]
Another PR (https://github.com/greenplum-db/gpdb/pull/13369) caused this one to fail with a NULL pointer dereference. I've added an appropriate check (the behavior of this function matches similar functions in the same file, so I don't believe this introduces any correctness issues). I've re-based the PR against `6X_STABLE` and added this check.

**Review by l-wang — APPROVED** [review]
+1 once pipeline is green.

**Review by dgkimura — COMMENTED** [review]
I have a couple questions since this doesn't look like a clean cherry-pick.

File: src/test/regress/sql/rpt.sql [hunk 1]
```diff
@@ -417,23 +418,80 @@ explain (costs off) create temp table tmp as select * from pg_class c join rtbl
...
-reset optimizer;
 reset enable_bitmapscan;
 
--- Github Issue 13532
-create table t1_13532(a int, b int) distributed replicated;
-create table t2_13532(a int, b int) distributed replicated;
-create index idx_t2_13532 on t2_13532(b);
-explain (costs off) select * from t1_13532 x, t2_13532 y where y.a < random() and x.b = y.b;
-set enable_bitmapscan = off;
-explain (costs off) select * from t1_13532 x, t2_13532 y where y.a < random() and x.b = y.b;
```
Comment (dgkimura) [inline src/test/regress/sql/rpt.sql:436]: Also looks like you deleted a test.  Is that intentional and if so why?

File: src/backend/commands/createas.c [hunk 2]
```diff
@@ -714,6 +714,9 @@ intorel_receive(TupleTableSlot *slot, DestReceiver *self)
 	DR_intorel *myState = (DR_intorel *) self;
 	Relation    into_rel = myState->rel;
 
+	if (into_rel == NULL)
+		return;
+
```
[the remark this answers is not in the document]
Comment (thedanhoffman) [reply src/backend/commands/createas.c:719]: This was a fix for a regression introduced by a merged PR into `6X_STABLE` since opening this PR originally.

File: src/test/regress/sql/rpt.sql (hunk 1 above)
Comment (thedanhoffman) [reply src/test/regress/sql/rpt.sql:436]: This was unintentional. Picked the wrong side of a merge conflict. Resolving it now

File: src/backend/commands/createas.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (thedanhoffman) [reply src/backend/commands/createas.c:719]: I agree. The only sign that the PR in question was broken was when I ran the tests included in this PR. The PR was merged after passing all existing tests, so it is possible that there wasn't adequate coverage. @SmartKeyerror any thoughts?

[the remark this answers is not in the document]
Comment (SmartKeyerror) [reply src/backend/commands/createas.c:719]: I can't think of a case where `into_rel` is null for the time being, and when I delete the judgment, the full pipeline doesn't fail in any case. So maybe we could delete it at this time?

[the remark this answers is not in the document]
Comment (thedanhoffman) [reply src/backend/commands/createas.c]: I can't reproduce the NULL pointer dereference when I've removed this change, so I went ahead and removed it. I'm unsure why it failed the first time, but if the pipeline is green with all of the current test cases (which includes the test which originally caused this to fail), then I'm comfortable calling it a one-off and merging this PR without the NULL pointer check.

File: src/test/regress/expected/rpt_optimizer.out [hunk 3]
```diff
@@ -702,71 +702,69 @@ DROP TABLE foopart;
...
    ->  Nested Loop
+         Join Filter: true
          ->  Seq Scan on t_hashdist
-         ->  Materialize
-               ->  Broadcast Motion 1:3  (slice1; segments: 1)
-                     ->  Result
-                           ->  Seq Scan on t_replicate_volatile
-                                 Filter: ((a)::double precision > random())
- Optimizer: Postgres query optimizer
-(9 rows)
```
Comment (dgkimura) [inline src/test/regress/expected/rpt_optimizer.out:722]: I feel like all the plan changes in this file are unrelated to your original change?  Is it intentional that you turned the GUCs off here, and if so why?

In this particular case, it looks like ORCA plan produces non-deterministic results, which I think is what the above FIXME is for?

File: src/backend/commands/createas.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (dgkimura) [reply src/backend/commands/createas.c]: Given that none your tests even do a CTAS, what you say sounds reasonable to me. Thanks for looking into it.

File: src/test/regress/expected/rpt_optimizer.out (hunk 3 above)
Comment (thedanhoffman) [reply src/test/regress/expected/rpt_optimizer.out:722]: `6X_STABLE` had no `rpt_optimizer.out` test file at the time this PR was made, and the current conflicts are against a previous PR merged (after this PR has been opened) and my current PR. I'll use the `6X_STABLE` version of all changes not relevant to this issue and follow up here

File: src/backend/commands/createas.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (thedanhoffman) [reply src/backend/commands/createas.c]: No problem
